cmd/block-generator/runner/run.go

```diff
@@ -0,0 +1,260 @@
+package runner
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	// Load the postgres sql.DB implementation
+	_ "github.com/lib/pq"
+
+	"github.com/algorand/indexer/cmd/block-generator/generator"
+	"github.com/algorand/indexer/util"
+)
+
+// Args are all the things needed to run a performance test.
+type Args struct {
+	// Path is a directory when passed to RunBatch, otherwise a file path.
+	Path                     string
+	IndexerBinary            string
+	IndexerPort              uint64
+	PostgresConnectionString string
+	RunDuration              time.Duration
+	ReportDirectory          string
+
+	indexerPort uint64
+}
+
+// Run is a public helper run the tests.
+// The test will run against the generator configuration file specified by 'args.Path'.
+// If 'args.Path' is a directory it should contain generator configuration files, a test will run using each file.
+func Run(args Args) error {
+	if _, err := os.Stat(args.ReportDirectory); !os.IsNotExist(err) {
+		return fmt.Errorf("report directory '%s' already exists", args.ReportDirectory)
+	}
+	os.Mkdir(args.ReportDirectory, os.ModeDir|os.ModePerm)
+
+	// Batch mode
+	pathStat, err := os.Stat(args.Path)
+	util.MaybeFail(err, "Unable to check path.")
+	if pathStat.IsDir() {
+		return filepath.Walk(args.Path, func(path string, info os.FileInfo, err error) error {
+			// Ignore the directory
+			if info.IsDir() {
+				return nil
+			}
+			runnerArgs := args
+			runnerArgs.Path = path
+			fmt.Printf("Running test for configuration '%s'\n", path)
+			return runnerArgs.run()
+		})
+	}
+
+	// Single file mode
+	return args.run()
+}
+
+func (r *Args) run() error {
+	// Start services
+	algodNet := fmt.Sprintf("localhost:%d", 11112)
+	indexerNet := fmt.Sprintf("localhost:%d", r.IndexerPort)
+	generatorShutdownFunc := startGenerator(r.Path, algodNet)
+	indexerShutdownFunc, err := startIndexer(r.IndexerBinary, algodNet, indexerNet, r.PostgresConnectionString)
+	if err != nil {
+		return fmt.Errorf("failed to start indexer: %w", err)
+	}
+
+	// Run the test, collecting results.
+	r.runTest(indexerNet, algodNet)
+
+	// Shutdown generator.
+	if err := generatorShutdownFunc(); err != nil {
+		return err
+	}
+
+	// Shutdown indexer
+	if err := indexerShutdownFunc(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Helper to record the import rate.
+func recordDataToFile(entry Entry, key string, out *os.File) error {
+	sum := 0.0
```

Comment:
Could you add units to the variable name? Maybe `totalSeconds`?

Reply:
The units are in the key, this function just divides the two numbers.